packages/flutter/lib/src/widgets/sensitive_content.dart

```diff
@@ -0,0 +1,373 @@
+/// [SensitiveContent] widgets in the tree will determine whether or not the
+/// screen will be obscured during media projection, e.g. screen sharing.
+///
+/// {@macro flutter.services.ContentSensitivity}
+///
+/// Currently, this widget is only supported on Android API 35+. On all lower Android
+/// versions and non-Android platforms, this does nothing; the screen will never be
+/// obscured regardless of the [sensitivity] set. To programmatically check if
+/// a device supports this widget, call [SensitiveContentService.isSupported].
+///
+/// See also:
+///
+///  * [ContentSensitivity], which are the different content sensitivity that a
+///    [SensitiveContent] widget can set.
+///
+/// This widget is marked visible for testing because it is not ready for use. It
+/// will be ready for use when it is implemented such that we can guarantee that no
+/// frames will be dropped when content is marked sensitive, ensuring that no sensitive
+/// content will be revealed during media projection.
+// TODO(camsim99): Fix `SensitiveContent` implementation to prevent revealing sensitive
+// content during media projection. Then, export this file to make the widget available
+// for use. See https://github.com/flutter/flutter/issues/160050 and
+// https://github.com/flutter/flutter/issues/164820.
+@visibleForTesting
+class SensitiveContent extends StatefulWidget {
+  /// Creates a [SensitiveContent] widget.
+  const SensitiveContent({super.key, required this.sensitivity, required this.child});
+
+  /// The sensitivity that the [SensitiveContent] widget should sets for the
+  /// Android native `View` hosting the widget tree.
+  final ContentSensitivity sensitivity;
+
+  /// The child widget of this [SensitiveContent].
+  ///
+  /// If the [sensitivity] is set to [ContentSensitivity.sensitive], then
+  /// the entire screen will be obscured when the screen is projected irrespective
+  /// to the parent/child widgets.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  @override
+  State<SensitiveContent> createState() => _SensitiveContentState();
+}
+
+class _SensitiveContentState extends State<SensitiveContent> {
+  Future<void> _sensitiveContentRegistrationFuture = Future<void>.value();
+  Future<void> _sensitiveContentUnregistrationFuture = Future<void>.value();
+
+  @override
+  void initState() {
+    super.initState();
+    _sensitiveContentRegistrationFuture = SensitiveContentHost.register(widget.sensitivity);
+  }
+
+  @override
+  void dispose() {
+    SensitiveContentHost.unregister(widget.sensitivity);
+    super.dispose();
+  }
+
+  @override
+  void didUpdateWidget(SensitiveContent oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    if (widget.sensitivity == oldWidget.sensitivity) {
+      return;
+    }
+
+    // Re-register SensitiveContent widget if the sensitivity changed.
+    _sensitiveContentRegistrationFuture = SensitiveContentHost.register(widget.sensitivity);
```

Comment:
we should have one method to replace the new old registration with new one instead of two. having two separate setSenstiveContent call worries me a bit.

Reply:
this way, you only need to keep track of one future

Reply:
I actually did this before but I thought it was breaking something. It was not; fixed.